File: go-server/internal/tools/post.go
// Package tools provides MCP tool implementations
package tools

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
	"time"

	"github.com/oyin-bo/autoreply/go-server/internal/auth"
	"github.com/oyin-bo/autoreply/go-server/internal/mcp"
	"github.com/oyin-bo/autoreply/go-server/pkg/errors"
)

// PostTool implements the post tool for creating posts and replies
type PostTool struct {
	credStore *auth.CredentialStore
	client    *http.Client
}

// NewPostTool creates a new post tool
func NewPostTool() (*PostTool, error) {
	credStore, err := auth.NewCredentialStore()
	if err != nil {
		return nil, fmt.Errorf("failed to initialize credential store: %w", err)
	}

	return &PostTool{
		credStore: credStore,
		client: &http.Client{
			Timeout: 30 * time.Second,
		},
	}, nil
}

// Name returns the tool name
func (t *PostTool) Name() string {
	return "post"
}

// Description returns the tool description
func (t *PostTool) Description() string {
	return "Create new posts with text content and optional reply functionality. Supports posting as authenticated users."
}

// InputSchema returns the JSON schema for tool input
func (t *PostTool) InputSchema() mcp.InputSchema {
	return mcp.InputSchema{
		Type: "object",
		Properties: map[string]mcp.PropertySchema{
			"postAs": {
				Type:        "string",
				Description: "Handle or DID to post as (uses default account if not specified)",
			},
			"text": {
				Type:        "string",
				Description: "Text content of the post (required)",
			},
			"replyTo": {
				Type:        "string",
				Description: "Post URI (at://...) or Bluesky URL (https://bsky.app/...) to reply to (optional)",
			},
		},
		Required: []string{"text"},
	}
}

// Call executes the post tool
func (t *PostTool) Call(ctx context.Context, args map[string]interface{}, _ *mcp.Server) (*mcp.ToolResult, error) {
	// Extract text parameter (required)
	textRaw, ok := args["text"]
	if !ok {
		return nil, errors.NewMCPError(errors.InvalidInput, "text parameter is required")
	}

	text, ok := textRaw.(string)
	if !ok {
		return nil, errors.NewMCPError(errors.InvalidInput, "text must be a string")
	}

	text = strings.TrimSpace(text)
	if text == "" {
		return nil, errors.NewMCPError(errors.InvalidInput, "text cannot be empty")
	}

	// Extract postAs parameter (optional, uses default if not provided)
	var postAs string
	if postAsRaw, ok := args["postAs"]; ok {
		if postAsStr, ok := postAsRaw.(string); ok {
			postAs = strings.TrimSpace(strings.TrimPrefix(postAsStr, "@"))
		}
	}

	// Get credentials for the user
	creds, err := t.getCredentials(postAs)
	if err != nil {
		return nil, err
	}

	// Extract replyTo parameter (optional)
	var replyTo string
	if replyToRaw, ok := args["replyTo"]; ok {
		if replyToStr, ok := replyToRaw.(string); ok {
			replyTo = strings.TrimSpace(replyToStr)
		}
	}

	// Create the post
	result, err := t.createPost(ctx, creds, text, replyTo)
	if err != nil {
		return nil, err
	}

	return result, nil
}

// getCredentials retrieves credentials for a user (by handle or DID)
func (t *PostTool) getCredentials(postAs string) (*auth.Credentials, error) {
	// If no postAs specified, use default
	if postAs == "" {
		defaultHandle, err := t.credStore.GetDefault()
		if err != nil {
			return nil, errors.NewMCPError(errors.InvalidInput, "No default account set. Please login first or specify postAs parameter.")
		}
		postAs = defaultHandle
	}

	// Load credentials
	creds, err := t.credStore.Load(postAs)
	if err != nil {
		return nil, errors.Wrap(err, errors.NotFound, fmt.Sprintf("No credentials found for %s. Please login first.", postAs))
	}

	return creds, nil
}

// createPost creates a new post, optionally as a reply
func (t *PostTool) createPost(ctx context.Context, creds *auth.Credentials, text, replyTo string) (*mcp.ToolResult, error) {
	// Build the post record
	record := map[string]interface{}{
		"$type":     "app.bsky.feed.post",
		"text":      text,
		"createdAt": time.Now().UTC().Format(time.RFC3339),
	}

	var replyInfo *ReplyInfo

	// If this is a reply, fetch the parent post
	if replyTo != "" {
		var err error
		replyInfo, err = t.getReplyInfo(ctx, creds, replyTo)
		if err != nil {
			return nil, errors.Wrap(err, errors.InvalidInput, "Failed to process replyTo parameter")
		}

		record["reply"] = map[string]interface{}{
			"root":   replyInfo.Root,
			"parent": replyInfo.Parent,
		}
	}

	// Create the record via AT Protocol API
	endpoint := "https://bsky.social/xrpc/com.atproto.repo.createRecord"

	requestBody := map[string]interface{}{
		"repo":       creds.DID,
		"collection": "app.bsky.feed.post",
		"record":     record,
	}

	jsonData, err := json.Marshal(requestBody)
	if err != nil {
		return nil, errors.Wrap(err, errors.InternalError, "Failed to marshal request")
	}

	req, err := http.NewRequestWithContext(ctx, "POST", endpoint, bytes.NewBuffer(jsonData))
	if err != nil {
		return nil, errors.Wrap(err, errors.InternalError, "Failed to create request")
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", creds.AccessToken))
	req.Header.Set("User-Agent", "autoreply/1.0")

	resp, err := t.client.Do(req)
	if err != nil {
		return nil, errors.Wrap(err, errors.InternalError, "Failed to create post")
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		var errorResp map[string]interface{}
		json.NewDecoder(resp.Body).Decode(&errorResp)
		return nil, errors.NewMCPError(errors.InternalError, fmt.Sprintf("Post creation failed with status %d: %v", resp.StatusCode, errorResp))
	}

	var createResp struct {
		URI string `json:"uri"`
		CID string `json:"cid"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&createResp); err != nil {
		return nil, errors.Wrap(err, errors.InternalError, "Failed to parse response")
	}

	// Build response message
	var message strings.Builder
	if replyInfo != nil {
		message.WriteString(fmt.Sprintf("# Reply Posted\n\n**Reply to:** %s\n\n**Your reply:**\n%s\n\n**Post URI:** `%s`\n",
			replyInfo.ParentText, text, createResp.URI))
	} else {
		message.WriteString(fmt.Sprintf("# Post Created\n\n**Text:**\n%s\n\n**Post URI:** `%s`\n",
			text, createResp.URI))
	}

	// Convert URI to Bluesky web URL
	webURL := t.atURIToBskyURL(createResp.URI, creds.Handle)
	message.WriteString(fmt.Sprintf("\n**View on Bluesky:** %s\n", webURL))

	return &mcp.ToolResult{
		Content: []mcp.ContentItem{
			{
				Type: "text",
				Text: message.String(),
			},
		},
	}, nil
}

// ReplyInfo contains information about a reply's parent and root posts
type ReplyInfo struct {
	Root       map[string]interface{}
	Parent     map[string]interface{}
	ParentText string
}

// getReplyInfo fetches information about the post being replied to
func (t *PostTool) getReplyInfo(ctx context.Context, creds *auth.Credentials, replyTo string) (*ReplyInfo, error) {
	// Parse the URI/URL
	postRef, err := parsePostReference(replyTo)
	if err != nil {
		return nil, err
	}

	// Fetch the post record
	endpoint := fmt.Sprintf("https://bsky.social/xrpc/com.atproto.repo.getRecord?repo=%s&collection=app.bsky.feed.post&rkey=%s",
		postRef.DID, postRef.RKey)

	req, err := http.NewRequestWithContext(ctx, "GET", endpoint, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", creds.AccessToken))
	req.Header.Set("User-Agent", "autoreply/1.0")

	resp, err := t.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch post: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		var errorResp map[string]interface{}
		json.NewDecoder(resp.Body).Decode(&errorResp)
		return nil, fmt.Errorf("failed to fetch post with status %d: %v", resp.StatusCode, errorResp)
	}

	var recordResp struct {
		URI   string                 `json:"uri"`
		CID   string                 `json:"cid"`
		Value map[string]interface{} `json:"value"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&recordResp); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	// Build reply structure
	replyInfo := &ReplyInfo{
		Parent: map[string]interface{}{
			"uri": recordResp.URI,
			"cid": recordResp.CID,
		},
	}

	// Extract parent text for display
	if text, ok := recordResp.Value["text"].(string); ok {
		replyInfo.ParentText = text
	}

	// Check if parent is already a reply - if so, use its root
	if parentReply, ok := recordResp.Value["reply"].(map[string]interface{}); ok {
		if root, ok := parentReply["root"].(map[string]interface{}); ok {
			replyInfo.Root = root
		} else {
			// Parent is a reply but has no root - use parent as root
			replyInfo.Root = replyInfo.Parent
		}
	} else {
		// Parent is not a reply - it becomes the root
		replyInfo.Root = replyInfo.Parent
	}

	return replyInfo, nil
}

// atURIToBskyURL converts an AT URI to a Bluesky web URL
func (t *PostTool) atURIToBskyURL(atURI, handle string) string {
	// Parse AT URI: at://{did}/{collection}/{rkey}
	if !strings.HasPrefix(atURI, "at://") {
		return atURI
	}

	parts := strings.Split(strings.TrimPrefix(atURI, "at://"), "/")
	if len(parts) < 3 {
		return atURI
	}

	rkey := parts[2]
	profile := strings.TrimPrefix(handle, "@")
	if profile == "" {
		profile = parts[0] // Use DID as fallback
	}

	return fmt.Sprintf("https://bsky.app/profile/%s/post/%s", profile, rkey)
}

File: rust-server/src/tools/post.rs
//! Post tool implementation
//!
//! Implements the `post(postAs, text, replyTo)` MCP tool

use crate::auth::storage::CredentialStorage;
use crate::auth::SessionManager;
use crate::cli::PostArgs;
use crate::error::AppError;
use crate::mcp::{McpResponse, ToolResult};
use anyhow::Result;
use serde_json::Value;
use tokio::time::{timeout, Duration};
use tracing::debug;

/// Handle post tool call
pub async fn handle_post(id: Option<Value>, args: Value) -> McpResponse {
    // Set total timeout to 120 seconds
    match timeout(Duration::from_secs(120), handle_post_impl(args)).await {
        Ok(result) => match result {
            Ok(content) => McpResponse::success(id, serde_json::to_value(content).unwrap()),
            Err(e) => McpResponse::error(id, e.error_code(), &e.message()),
        },
        Err(_) => McpResponse::error(id, "timeout", "Post request exceeded 120 second timeout"),
    }
}

async fn handle_post_impl(args: Value) -> Result<ToolResult, AppError> {
    // Parse arguments
    let post_args: PostArgs = serde_json::from_value(args)
        .map_err(|e| AppError::InvalidInput(format!("Invalid arguments: {}", e)))?;

    // Execute using shared implementation
    execute_post(post_args).await
}

/// Execute post tool (shared implementation for MCP and CLI)
pub async fn execute_post(post_args: PostArgs) -> Result<ToolResult, AppError> {
    debug!(
        "Post request for account: {}, text: '{}'",
        post_args.postAs, post_args.text
    );

    // Get credentials for the account
    let storage = CredentialStorage::new()?;

    // Try to get stored session first (for OAuth accounts)
    let session = if let Some(stored_session) = storage.get_session(&post_args.postAs)? {
        debug!("Using stored session for {}", post_args.postAs);
        stored_session
    } else {
        // Fallback to creating new session with credentials (for app password accounts)
        debug!(
            "No stored session, creating new session for {}",
            post_args.postAs
        );
        let credentials = storage.get_credentials(&post_args.postAs)?;
        let session_manager = SessionManager::new()?;
        session_manager.login(&credentials).await?
    };

    debug!("Authenticated as {} (DID: {})", session.handle, session.did);

    // Parse reply-to if provided
    let reply_ref = if let Some(reply_to) = &post_args.replyTo {
        Some(parse_and_fetch_reply(&session, reply_to).await?)
    } else {
        None
    };

    // Create the post
    let client = crate::http::client_with_timeout(std::time::Duration::from_secs(120));
    let url = format!("{}/xrpc/com.atproto.repo.createRecord", session.service);

    let mut record = serde_json::json!({
        "$type": "app.bsky.feed.post",
        "text": post_args.text,
        "createdAt": chrono::Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true),
    });

    // Add reply information if present
    if let Some(reply) = reply_ref {
        record["reply"] = reply;
    }

    let body = serde_json::json!({
        "repo": session.did,
        "collection": "app.bsky.feed.post",
        "record": record,
    });

    debug!("Creating post with body: {}", serde_json::to_string(&body)?);

    let response = client
        .post(&url)
        .header("Authorization", format!("Bearer {}", session.access_jwt))
        .json(&body)
        .send()
        .await
        .map_err(|e| AppError::NetworkError(format!("Post creation request failed: {}", e)))?;

    let status = response.status();
    if !status.is_success() {
        let error_text = response
            .text()
            .await
            .unwrap_or_else(|_| "Unknown error".to_string());
        return Err(AppError::NetworkError(format!(
            "Post creation failed with status {}: {}",
            status, error_text
        )));
    }

    let result: serde_json::Value = response
        .json()
        .await
        .map_err(|e| AppError::ParseError(format!("Failed to parse response: {}", e)))?;

    let post_uri = result["uri"]
        .as_str()
        .ok_or_else(|| AppError::ParseError("No URI in response".to_string()))?;

    debug!("Post created successfully: {}", post_uri);

    // Format result as markdown
    let markdown = if post_args.replyTo.is_some() {
        format!(
            "# Reply Posted\n\n**Post URI:** {}\n\n**Text:** {}\n\n**Reply To:** {}\n",
            post_uri,
            post_args.text,
            post_args.replyTo.as_ref().unwrap()
        )
    } else {
        format!(
            "# Post Created\n\n**Post URI:** {}\n\n**Text:** {}\n",
            post_uri, post_args.text
        )
    };

    Ok(ToolResult::text(markdown))
}

/// Parse a post URI/URL and fetch the post details to create a reply reference
async fn parse_and_fetch_reply(
    session: &crate::auth::Session,
    reply_to: &str,
) -> Result<serde_json::Value, AppError> {
    // Parse the URI using the shared utility
    let post_ref = crate::bluesky::uri::parse_post_uri(reply_to).await?;

    // Fetch the post to get its CID
    let client = crate::http::client_with_timeout(std::time::Duration::from_secs(120));
    let url = format!(
        "{}/xrpc/com.atproto.repo.getRecord?repo={}&collection=app.bsky.feed.post&rkey={}",
        session.service, post_ref.did, post_ref.rkey
    );

    debug!("Fetching reply-to post from: {}", url);

    let response = client
        .get(&url)
        .header("Authorization", format!("Bearer {}", session.access_jwt))
        .send()
        .await
        .map_err(|e| AppError::NetworkError(format!("Failed to fetch reply-to post: {}", e)))?;

    let status = response.status();
    if !status.is_success() {
        let error_text = response
            .text()
            .await
            .unwrap_or_else(|_| "Unknown error".to_string());
        return Err(AppError::NetworkError(format!(
            "Failed to fetch reply-to post with status {}: {}",
            status, error_text
        )));
    }

    let post_data: serde_json::Value = response
        .json()
        .await
        .map_err(|e| AppError::ParseError(format!("Failed to parse post data: {}", e)))?;

    let post_uri = post_data["uri"]
        .as_str()
        .ok_or_else(|| AppError::ParseError("No URI in post data".to_string()))?;
    let post_cid = post_data["cid"]
        .as_str()
        .ok_or_else(|| AppError::ParseError("No CID in post data".to_string()))?;

    // Check if the post itself is a reply to build the proper reply chain
    let root_ref = if let Some(reply) = post_data["value"].get("reply") {
        if let Some(root) = reply.get("root") {
            root.clone()
        } else {
            serde_json::json!({
                "uri": post_uri,
                "cid": post_cid
            })
        }
    } else {
        serde_json::json!({
            "uri": post_uri,
            "cid": post_cid
        })
    };

    Ok(serde_json::json!({
        "root": root_ref,
        "parent": {
            "uri": post_uri,
            "cid": post_cid
        }
    }))
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[tokio::test]
    async fn test_post_args_parsing() {
        let args = json!({
            "postAs": "test.bsky.social",
            "text": "Hello, world!"
        });

        let parsed: PostArgs = serde_json::from_value(args).unwrap();
        assert_eq!(parsed.postAs, "test.bsky.social");
        assert_eq!(parsed.text, "Hello, world!");
        assert!(parsed.replyTo.is_none());
    }

    #[tokio::test]
    async fn test_post_args_with_reply() {
        let args = json!({
            "postAs": "test.bsky.social",
            "text": "Reply text",
            "replyTo": "at://did:plc:abc/app.bsky.feed.post/123"
        });

        let parsed: PostArgs = serde_json::from_value(args).unwrap();
        assert_eq!(parsed.postAs, "test.bsky.social");
        assert_eq!(parsed.text, "Reply text");
        assert_eq!(
            parsed.replyTo,
            Some("at://did:plc:abc/app.bsky.feed.post/123".to_string())
        );
    }
}
